File: strongtify_client/src/types/song.ts
import { Language } from "@/libs/enums";
import { Artist } from "./artist";
import { PagingQuery } from "./paging";
import { Genre } from "./genre";

export type Song = {
    id: string;
    name: string;
    alias: string;
    songUrl?: string;
    imageUrl?: string;
    length: number;
    releasedAt?: string;
    likeCount: number;
    listenCount: number;
    language?: Language;
    artists?: Artist[];
}

export type TopSong = Song & {
    listenCountInTimeRange: number;
}

export type SongDetail = {
    id: string;
    name: string;
    alias: string;
    songUrl?: string;
    imageUrl?: string;
    length: number;
    releasedAt?: string;
    likeCount: number;
    listenCount: number;
    language?: Language;
    artists?: Artist[];
    genres?: Genre[];
}

export type CudSongResponse = {
    id: string;
    name: string;
    alias: string;
    songUrl?: string;
    imageUrl?: string;
    length: number;
    releasedAt?: Date;
    likeCount: number;
    listenCount: number;
    language?: Language;
}

export type CreateUpdateSongRequest = {
    name: string;
    length?: number;
    releasedAt?: string;
    songUrl?: string;
    language: Language;
    artistIds?: string[];
    genreIds?: string[];
    image?: File;
}

export type SongQuery = PagingQuery & {
    language?: Language;
    artistId?: string;
    genreId?: string;
}

File: strongtify_mobile_app/lib/models/song/song.dart
import 'package:strongtify_mobile_app/models/artist/artist.dart';

class Song {
  Song(
      {required this.id,
      required this.name,
      required this.alias,
      this.imageUrl,
      this.songUrl,
      this.likeCount = 0,
      this.listenCount = 0,
      this.length = 0,
      this.language = 'NONE',
      this.releasedAt,
      this.artists});

  final String id;
  final String name;
  final String alias;
  final String? imageUrl;
  final String? songUrl;
  final int likeCount;
  final int listenCount;
  final int length;
  final String language;
  final DateTime? releasedAt;
  final List<Artist>? artists;

  factory Song.fromMap(Map<String, dynamic> data) {
    return Song(
      id: data['id'],
      name: data['name'],
      alias: data['alias'],
      imageUrl: data['imageUrl'],
      songUrl: data['songUrl'],
      likeCount: data['likeCount'] ?? 0,
      listenCount: data['listenCount'] ?? 0,
      length: data['length'] ?? 0,
      language: data['language'],
      releasedAt:
          data['releasedAt'] != null ? DateTime.parse(data['releasedAt']) : null,
      artists: data['artists'] != null
          ? (data['artists'] as List).map((e) => Artist.fromMap(e)).toList()
          : null,
    );
  }
}
